File: backend/src/main/java/com/banco/api/service/MovimientoService.java
package com.banco.api.service;

import com.banco.api.entity.Movimiento;
import java.util.List;

/**
 * Interfaz de servicio para gestionar Movimientos Financieros (Transacciones).
 * Define operaciones para depósitos, retiros e historial de transacciones.
 */
public interface MovimientoService {

    /**
     * Registra un nuevo movimiento (transacción).
     * Impacta el saldo de la cuenta.
     * 
     * @param movimiento - Los detalles del movimiento (cuenta, monto, tipo).
     * @return El movimiento creado con saldo actualizado.
     */
    Movimiento createMovimiento(Movimiento movimiento);

    /**
     * Actualiza un movimiento.
     * NOTA: Esta operación está restringida por razones de seguridad.
     * 
     * @param movimientoDTO - Los datos del movimiento.
     * @return El movimiento actualizado.
     */
    Movimiento updateMovimiento(com.banco.api.dto.MovimientoDTO movimientoDTO);

    /**
     * Elimina un movimiento y revierte su impacto en el saldo de la cuenta.
     * 
     * @param deleteDTO - El DTO con el ID del movimiento a eliminar.
     */
    void deleteMovimiento(com.banco.api.dto.DeleteDTO deleteDTO);

    /**
     * Recupera todos los movimientos.
     * 
     * @return Lista de todos los movimientos.
     */
    List<Movimiento> getAllMovimientos();
}

File: frontend/src/services/movimientoService.ts
import { api } from './api';
import type { Movimiento, BaseResponse } from '../types';

/**
 * Obtiene todos los movimientos de la API.
 * @returns Una promesa que resuelve a un array de objetos Movimiento.
 */
export const getMovimientos = async () => {
    const response = await api.get<BaseResponse<Movimiento[]>>('/movimientos');
    return response.data.data;
};

/**
 * Crea un nuevo movimiento.
 * @param movimiento - Los datos del movimiento a crear.
 * @returns Una promesa que resuelve al objeto Movimiento creado.
 */
export const createMovimiento = async (movimiento: Movimiento) => {
    const response = await api.post<BaseResponse<Movimiento>>('/movimientos', movimiento);
    return response.data.data;
};

/**
 * Actualiza un movimiento existente.
 * @param id - El ID del movimiento a actualizar.
 * @param movimiento - Los datos actualizados del movimiento.
 * @returns Una promesa que resuelve al objeto Movimiento actualizado.
 */
export const updateMovimiento = async (id: number, movimiento: Movimiento) => {
    const response = await api.put<BaseResponse<Movimiento>>(`/movimientos/${id}`, movimiento);
    return response.data.data;
};

/**
 * Elimina un movimiento por ID.
 * @param id - El ID del movimiento a eliminar.
 * @returns Una promesa que resuelve cuando la eliminación se completa.
 */
export const deleteMovimiento = async (id: number) => {
    return api.delete(`/movimientos/${id}`);
};
